src/tagmanager/tm_ctags.c

```diff
@@ -271,12 +274,152 @@ void tm_ctags_clear_ignore_symbols(void)
 }
 
 
+/* call after all tags have been collected so we don't have to handle reparses
+ * with the counter (which gets complicated when also subparsers are involved) */
+static void rename_anon_tags(TMSourceFile *source_file)
+{
+	gchar *anon_counter_table = NULL;
+	GPtrArray *removed_typedefs = NULL;
+	guint i;
+
+	for (i = 0; i < source_file->tags_array->len; i++)
+	{
+		TMTag *tag = TM_TAG(source_file->tags_array->pdata[i]);
+		if (tm_tag_is_anon(tag))
+		{
+			gchar *orig_name, *new_name = NULL;
+			guint j;
+			guint new_name_len, orig_name_len;
+			guint scope_len = tag->scope ? strlen(tag->scope) : 0;
+			gchar kind = tag->kind_letter;
+
+			orig_name = tag->name;
+			orig_name_len = strlen(orig_name);
+
+			if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
+			{
+				/* First check if there's a typedef behind the scope nesting
+				 * such as typedef struct {} Foo; - in this case we can replace
+				 * the anon tag with Foo */
+				for (j = i + 1; j < source_file->tags_array->len; j++)
+				{
+					TMTag *nested_tag = TM_TAG(source_file->tags_array->pdata[j]);
+					guint nested_scope_len = nested_tag->scope ? strlen(nested_tag->scope) : 0;
+
+					/* Nested tags have longer scope than the parent - once the scope
+					 * is equal or lower than the parent scope, we are outside the tag's
+					 * scope. */
+					if (nested_scope_len <= scope_len)
+						break;
+				}
+
+				/* We are out of the nesting - the next tag could be a typedef */
+				if (j < source_file->tags_array->len)
+				{
+					TMTag *typedef_tag = TM_TAG(source_file->tags_array->pdata[j]);
+					guint typedef_scope_len = typedef_tag->scope ? strlen(typedef_tag->scope) : 0;
+
+					/* Should be at the same scope level as the anon tag */
+					if (typedef_tag->type == tm_tag_typedef_t &&
+						typedef_scope_len == scope_len &&
+						g_strcmp0(typedef_tag->var_type, tag->name) == 0)
+					{
+						/* set the name of the original anon tag and pretend
+						 * it wasn't a anon tag */
+						tag->name = g_strdup(typedef_tag->name);
+						tag->flags &= ~tm_tag_flag_anon_t;
+						new_name = tag->name;
+						/* the typedef tag will be removed */
+						if (!removed_typedefs)
+							removed_typedefs = g_ptr_array_new();
+						g_ptr_array_add(removed_typedefs, GUINT_TO_POINTER(j));
+					}
+				}
+			}
+
+			/* there's no typedef name for the anon tag so let's generate one  */
+			if (!new_name)
+			{
+				gchar buf[50];
+				guint anon_counter;
+				const gchar *kind_name = tm_ctags_get_kind_name(kind, tag->lang);
+
+				if (!anon_counter_table)
+					anon_counter_table = g_new0(gchar, 256);
```

Comment:
Err, well, I thought of simply placing the array on the stack (at the function begin). I'm also not sure if 8 bit counter will suit everyone. I think generated sources (from a large vala file or someting like that) could easily overflow 256 anon structs.

I think a `gint anon_counter_table[256]` should work.

Reply:
Yes, but then you have to zero it manually (which will be very fast but you will have to do it every time instead of lazily like now). No problem changing it but I think it's about the same like the current code.

The `gchar` is definitely my error, I got confused by that 256 value, will correct that.

Reply:
I don't feel strong about array vs. g_new0, do as you chose (hint: zero-initialization of arrays: `gint foo[N] = {0}`)